### ISSUE TITLE
JRuby build is slow

### ISSUE THREAD

**Issue by smellsblue** [bundler/gemstash#46]
The JRuby build is massively slower than MRI. I suspect this is due to JRuby warm up time happening in all the spawns in the integration specs. I've heard there are gems which can pre-warm the JVM... these may drastically speed up the build.

**Comment by smellsblue** [comment]
I added some timing to the execs happening in the integration specs. At first I thought I ran into something strange with some tests running extremely slow, except it ended up being the same situation as seen in [a build based on my temporary branch](https://travis-ci.org/bundler/gemstash/builds/88953593). That problem causes the build to run in excess of an hour and is unrelated to the normal build, which takes almost 5 minutes on JRuby. Interestingly, that same build took similar amount of time on ruby-head. I suspect it has something to do with the test servers being launched in the integration specs and the same test servers launched in a separate spec (a temporary one that was intended to be used to help solve the Bundler full index header issue). It could be excessively long due to the same port being in use, or something along those lines.

As for the normal build, here is the running times for a JRuby build on my system:
```
EXEC gem push --key test --host 'http://127.0.0.1:9042/private' '[redacted]/gemstash/spec/data/gems/speaker/pkg/speaker-0.1.0.gem' IN 4.554s
EXEC gem yank --key test 'speaker' --version 0.1.0 IN 2.779s
EXEC gem yank --key test 'speaker' --version 0.1.0 --undo IN 2.833s
EXEC bundle IN 3.618s
EXEC bundle exec speaker hi IN 3.901s
EXEC bundle --full-index IN 3.93s
EXEC bundle exec speaker hi IN 3.486s
EXEC bundle IN 3.476s
EXEC bundle exec speaker hi IN 3.682s
EXEC bundle --full-index IN 3.654s
EXEC bundle exec speaker hi IN 4.168s
EXEC bundle IN 3.205s
EXEC bundle exec speaker hi IN 3.644s
EXEC bundle IN 3.561s
EXEC bundle exec speaker hi IN 3.828s
EXEC bundle IN 3.726s
EXEC bundle exec speaker hi IN 4.019s
EXEC bundle --full-index IN 3.995s
EXEC bundle exec speaker hi IN 3.901s
EXEC bundle IN 3.522s
EXEC bundle exec speaker hi IN 3.96s
EXEC bundle --full-index IN 25.573s
EXEC bundle exec speaker hi IN 3.801s
EXEC bundle IN 29.156s
EXEC bundle exec speaker hi IN 4.08s
EXEC bundle --full-index IN 10.98s
EXEC bundle exec speaker hi IN 5.415s
EXEC bundle IN 13.744s
EXEC bundle exec speaker hi IN 5.088s
EXEC bundle --full-index IN 14.87s
EXEC bundle exec speaker hi IN 5.197s
EXEC bundle IN 11.742s
EXEC bundle exec speaker hi IN 4.961s
EXEC bundle --full-index IN 10.858s
EXEC bundle exec speaker hi IN 5.374s
EXEC bundle IN 8.463s
EXEC bundle exec speaker hi IN 4.996s
EXEC bundle --full-index IN 13.183s
EXEC bundle exec speaker hi IN 5.086s
```

The `bundle exec speaker hi` regularly takes 3-5 seconds and seems to be about the same time needed for JRuby to warm up.

Bundling varies much more drastically, and might need closer attention to see what is going on.

I've attempted various things to improve the situation, but nothing seems to give anywhere near the gain I'd like to see, but I will add the JRuby `--dev` option for the execs, which seems to make a useful dent.

**Comment by smellsblue** [comment]
I got things working with a bit smoother, but it wasn't a massive gain, and I think we can do better.

Inspired by a JRuby feature `launch.inproc=true`, I think exec_helpers could do an in-process exec in a thread rather than spawn a process. I tried using `launch.inproc=true`, except it won't work directly, as it blocks when native code is loaded, and then doesn't work even if you eliminate native code (which is especially hard because apparently the logger stdlib uses native code).

We could probably do it ourselves by launching a JRuby scriptlet, as long as we can override some environment variables, capture the exit code (also not let it actually exit the JVM), and pass arguments in.
